File: web_version/pmugg_files/network/edgeNet.js
ms.edgeNet = function() {
	this.halfEdges = [];
	this.primal = null;
	this.network = null;

	this.id = ms.counter.add('edgeNet');
};

ms.counter.register('edgeNet');

ms.edgeNet.prototype.getHalfEdges = function() {
	return this.halfEdges;
};

ms.edgeNet.prototype.getPrimal    = function() { return this.primal; }
ms.edgeNet.prototype.getNetwork   = function() { return this.network; }

ms.edgeNet.prototype.setPrimal = function(primal) {
	this.primal = primal;
};

ms.edgeNet.prototype.connectNet = function(network) {
	this.network = network;
	network.addEdge(this);
	return this;
};

ms.edgeNet.prototype.addHalfEdge = function(halfEdge, index) {
	if (!this.halfEdges[index]) {
		this.halfEdges[index] = [];
	}
	this.halfEdges[index].push(halfEdge);
};

ms.edgeNet.prototype.removeHalfEdge = function(halfEdge, index) {
	ms.remove(halfEdge, this.halfEdges[index]);
};

ms.edgeNet.prototype.copyConnection = function(copy) {
	var copyNet = copy.getNetwork();
	var self = this;
	this.halfEdges = copy.getHalfEdges().map((halfEdges) => {
		return halfEdges.map((half) => {
			return self.network.convertHalfEdge(copyNet, half);
		});
	});
};

ms.edgeNet.prototype.export = function() {
	var self = this;
	return {	
		halfEdges: this.getHalfEdges().map((halfEdges) => {
			return halfEdges.map((half) => {
				return self.network.halfEdgeIndex(half);
			});
		}),
	}
};

ms.edgeNet.prototype.import = function(json) {
	var self = this;
	this.halfEdges = json.halfEdges.map((halfEdges) => {
		return halfEdges.map((half) => {
			return self.network.getHalfEdges()[half];
		});
	});
};

// This is just a sanity check. This should be true if this is valid.
ms.edgeNet.prototype.inNetwork = function() {
	return this.network.getEdges().includes(this);
};

ms.edgeNet.prototype.highlight = function(view) {
	var halfEdges = [];
	this.halfEdges.forEach((halfArray) => {
		halfEdges = halfEdges.concat(halfArray);
	});
	var options = ms.halfEdgeSet.create(halfEdges);
	this.network.highlight(view, options);
};

ms.edgeNet.prototype.print = function() {
	var interior = this.primal.getInterior();
	if (interior != this) {
		console.log ('Boundary Edge');
		interior.print();
	} else {
		ms.highlight(this);
	}
};

ms.edgeNet.prototype.requiresShapeView = function () {
	return true;
};

// Merge edgeB into this edge. Assumes this is an interior edge.
ms.edgeNet.prototype.merge = function(edgeB, mergeForward) {
	var interior = this.network;

	var halfsB = edgeB.getHalfEdges();
	this.halfEdges.forEach((half, edgeIndex) => {
		var halfA = half[0];
		var halfB = halfsB[edgeIndex][0];
		var forward = halfA.getForward();
		if (!(forward ^ mergeForward)) {
			var nextB = halfB.getNext();
			halfA.disconnectHalfEdge();
			halfB.disconnectHalfEdge();
			halfA.connectHalfEdge(nextB);
		} else {
			var prevB = halfB.getPrev();
			if (prevB) {
				prevB.disconnectHalfEdge();
				halfB.disconnectHalfEdge();
				prevB.connectHalfEdge(halfA);
			}
			halfA.connectVertex(halfB.getVertex(), halfA.getVertexIndex());
		}		
		var face = halfB.getFace();
		face.replaceHalfEdge(halfB, halfA);
		interior.removeHalfEdge(halfB);
	});
	interior.removeEdge(edgeB);
};

File: cpp_version/pmugg/network/edge_net.cpp
#include "pch.h"
#include "edge_net.h"
#include "network.h"
#include "half_edge_net.h"
// #include "view.h"
#include "../util/util.h"

namespace ms {

int EdgeNet::nextId = 0;

EdgeNet::EdgeNet()
    : type(nullptr)
    , network(nullptr)
    , id(nextId++) {}

const std::vector<std::vector<HalfEdgeNet*>>& EdgeNet::getHalfEdges() const {
    return halfEdges;
}

//PrimalEdge* EdgeNet::getPrimal() const {
//    return primal;
//}

Network* EdgeNet::getNetwork() const {
    return network;
}

int EdgeNet::getId() const {
    return id;
}

//void EdgeNet::setPrimal(PrimalEdge* p) {
//    primal = p;
//}

EdgeNet* EdgeNet::connectNet(Network* net) {
    network = net;
    network->addEdge(this);
    return this;
}

void EdgeNet::addHalfEdge(HalfEdgeNet* halfEdge, int index) {
    if (index >= halfEdges.size()) {
        halfEdges.resize(index + 1);
    }
    halfEdges[index].push_back(halfEdge);
}

void EdgeNet::removeHalfEdge(HalfEdgeNet* halfEdge, int index) {
    if (index < halfEdges.size()) {
        halfEdges.erase(halfEdges.begin() + index);
    }
}

void EdgeNet::copyConnection(const EdgeNet* copy) {
    auto* copyNet = copy->getNetwork();
    auto copyHalfEdges = copy->getHalfEdges();
    
    halfEdges.clear();
    halfEdges.resize(copyHalfEdges.size());
    
    for (size_t i = 0; i < copyHalfEdges.size(); i++) {
        for (auto* half : copyHalfEdges[i]) {
            halfEdges[i].push_back(network->convertHalfEdge(copyNet, half));
        }
    }
}

bool EdgeNet::inNetwork() const {
    auto vec = network->getEdges();
    return std::find(vec.begin(), vec.end(), this) != vec.end();
}

//void EdgeNet::highlight(View* view) {
//    std::vector<HalfEdgeNet*> allHalfEdges;
//    for (const auto& halfArray : halfEdges) {
//        allHalfEdges.insert(allHalfEdges.end(), halfArray.begin(), halfArray.end());
//    }
//    
//    auto options = HalfEdgeSet::create(allHalfEdges);
//    network->highlight(view, options);
//}
//
//void EdgeNet::print() const {
//    auto* interior = primal->getInterior();
//    if (interior != this) {
//        std::cout << "Boundary Edge" << std::endl;
//        interior->print();
//    } else {
//        ms::highlight(this);
//    }
//}

void EdgeNet::merge(EdgeNet* edgeB, bool mergeForward) {
    auto* interior = network;

    auto halfsB = edgeB->getHalfEdges();
    for (size_t edgeIndex = 0; edgeIndex < halfEdges.size(); edgeIndex++) {
        auto* halfA = halfEdges[edgeIndex][0];
        auto* halfB = halfsB[edgeIndex][0];
        auto forward = halfA->getForward();
        
        if (!(forward ^ mergeForward)) {
            auto* nextB = halfB->getNext();
            halfA->disconnectHalfEdge();
            halfB->disconnectHalfEdge();
            halfA->connectHalfEdge(nextB);
        } else {
            auto* prevB = halfB->getPrev();
            if (prevB) {
                prevB->disconnectHalfEdge();
                halfB->disconnectHalfEdge();
                prevB->connectHalfEdge(halfA);
            }
            halfA->connectVertex(halfB->getVertex(), halfA->getVertexIndex());
        }
        
        auto* face = halfB->getFace();
        face->replaceHalfEdge(halfB, halfA, false);
        interior->removeHalfEdge(halfB);
    }
    interior->removeEdge(edgeB);
}

//Json EdgeNet::export() const {
//    Json json;
//    json["halfEdges"] = Json::array();
//    
//    for (const auto& halfArray : halfEdges) {
//        Json arrayJson = Json::array();
//        for (auto* half : halfArray) {
//            arrayJson.push_back(network->halfEdgeIndex(half));
//        }
//        json["halfEdges"].push_back(arrayJson);
//    }
//    
//    return json;
//}

void EdgeNet::import(const Json& json) {
    halfEdges.clear();
    auto& networkHalfEdges = network->getHalfEdges();
    
    if (json.contains("halfEdges")) {
        for (const auto& arrayJson : json["halfEdges"]) {
            std::vector<HalfEdgeNet*> halfArray;
            for (const auto& index : arrayJson) {
                halfArray.push_back(networkHalfEdges[index.get<int>()]);
            }
            halfEdges.push_back(halfArray);
        }
    }
}

} // namespace ms 